apps/manager_app/user_training.c

```diff
@@ -0,0 +1,285 @@
+ * or substantially, from the functionality of the Software. Any license
+ * notice or attribution required by the License must also include
+ * this Commons Clause License Condition notice.
+ *
+ * Software: All X1Wallet associated files.
+ * License: MIT
+ * Licensor: HODL TECH PTE LTD
+ *
+ ******************************************************************************
+ */
+
+/*****************************************************************************
+ * INCLUDES
+ *****************************************************************************/
+
+#include "events.h"
+#include "manager/get_device_info.pb.h"
+#include "manager/train_user.pb.h"
+#include "manager_api.h"
+#include "onboarding.h"
+#include "status_api.h"
+#include "ui_delay.h"
+#include "ui_joystick_training.h"
+
+/*****************************************************************************
+ * EXTERN VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE MACROS AND DEFINES
+ *****************************************************************************/
+
+#define JOYSTICK_TRAIN_STEPS 5
+
+#undef SEND_ERROR
+#define SEND_ERROR(a, b)                                                       \
+  fill_training_error(a, b);                                                   \
+  send_msg_to_host(a);
+
+/**
+ * @details The TRAINING_STEP macro takes message (m - string), user action (a -
+ * joystick_actions_t), context (c - training_context) and flow status update
+ * value (f). The macro **ignores timeout events** since currently, there is not
+ * way to disable it. Also, in case of abort, it will clear usb event
+ * (usb_clear_event) and silently return.
+ */
+#define TRAINING_STEP(m, a, c, f)                                              \
+  do {                                                                         \
+    joystick_train_init(m, a);                                                 \
+    do {                                                                       \
+      (c).events =                                                             \
+          get_events((c).evt_config.evt_selection, (c).evt_config.timeout);    \
+      if ((c).events.p0_event.abort_evt) {                                     \
+        usb_clear_event();                                                     \
+        return;                                                                \
+      }                                                                        \
+    } while (!(c).events.ui_event.event_occured);                              \
+    core_status_set_flow_status(f);                                            \
+  } while (0);
+
+/*****************************************************************************
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+
+typedef struct joystick_step {
```

Comment:
Can you please describe structure elements

Specifically status

Reply:
added comments